staging/src/k8s.io/endpointslice/util/controller_utils.go

```diff
@@ -47,22 +47,100 @@ var semanticIgnoreResourceVersion = conversion.EqualitiesOrDie(
 	},
 )
 
-// GetPodServiceMemberships returns a set of Service keys for Services that have
-// a selector matching the given pod.
-func GetPodServiceMemberships(serviceLister v1listers.ServiceLister, pod *v1.Pod) (sets.String, error) {
-	set := sets.String{}
-	services, err := serviceLister.Services(pod.Namespace).List(labels.Everything())
+// PodProjectionKey encapsulates all information required to find services the pod
+// maps to and if the pod changed in a way that may change the service endpoints.
+type PodProjectionKey struct {
+	Namespace  string
+	Labels     labels.Set
+	OldLabels  labels.Set
+	PodChanged bool
+}
+
+func GetPodProjectionKey(obj interface{}) *PodProjectionKey {
+	if pod := getPodFromObject(obj); pod != nil {
+		return &PodProjectionKey{
+			Namespace: pod.Namespace,
+			Labels:    labels.Set(pod.Labels),
+		}
+	}
+	return nil
+}
+
+func GetPodUpdateProjectionKey(oldObj, newObj interface{}) *PodProjectionKey {
+	newPod := getPodFromObject(newObj)
+	oldPod := getPodFromObject(oldObj)
+	if newPod == nil || oldPod == nil {
+		return nil
```

Comment:
I don't understand it. If only one of these is nil, some changes actually happened.

Reply:
I think I understood the pattern now, but it's quite misleading to me (see my comments above).

I would actually go with setting old/new as they really are (as stated in my other comments) and here basically do:
```
if newPod == nil && oldPod == nil {
   HandleError(...)
   return
}
if oldPod == nil {
  return PodProjectionKey(newPod)
}
if newPod == nil {
  return PodProjectionKey(oldPod)
}

// the rest of the logic as it is now
```

Thoughts?